File: frontend/web/types/teacher-mode.ts
/**
 * 教師モードの型定義
 */

/** 音声セグメントのタイプ */
export type AudioSegmentType = 'phrase' | 'translation' | 'explanation' | 'pause';

/** 再生状態 */
export type PlaybackStatus = 'stopped' | 'playing' | 'paused';

/** 音声品質 */
export type AudioQuality = 'standard' | 'premium';

/** 再生速度 */
export type PlaybackSpeed = 0.5 | 0.75 | 1.0 | 1.25 | 1.5 | 2.0;

/** リピート回数 */
export type RepeatCount = 1 | 2 | 3;

/** 教師モード設定 */
export interface TeacherModeSettings {
  /** 再生速度 */
  speed: PlaybackSpeed;
  /** ページ間隔（秒） */
  pageInterval: number;
  /** リピート回数 */
  repeatCount: RepeatCount;
  /** 音質 */
  audioQuality: AudioQuality;
  /** 学習内容 */
  content: {
    /** 母国語訳を含む */
    includeTranslation: boolean;
    /** 単語解説を含む */
    includeWordExplanation: boolean;
    /** 文法解説を含む */
    includeGrammarExplanation: boolean;
    /** 発音練習を含む */
    includePronunciationPractice: boolean;
    /** 例文を含む */
    includeExampleSentences: boolean;
  };
}

/** 音声セグメント */
export interface AudioSegment {
  /** ID */
  id: string;
  /** タイプ */
  type: AudioSegmentType;
  /** 音声URL */
  audioUrl: string;
  /** 長さ（ミリ秒） */
  duration: number;
  /** テキスト */
  text: string;
  /** 言語 */
  language: string;
}

/** ページ音声 */
export interface PageAudio {
  /** ページ番号 */
  pageNumber: number;
  /** セグメント */
  segments: AudioSegment[];
  /** 総長さ（ミリ秒） */
  totalDuration: number;
}

/** 教師モードプレイリスト */
export interface TeacherModePlaylist {
  /** プレイリストID */
  id: string;
  /** 書籍ID */
  bookId: string;
  /** ページ */
  pages: PageAudio[];
  /** 設定 */
  settings: TeacherModeSettings;
  /** 総長さ（ミリ秒） */
  totalDuration: number;
}

/** 再生状態 */
export interface PlaybackState {
  /** 状態 */
  status: PlaybackStatus;
  /** 現在のページ */
  currentPage: number;
  /** 現在のセグメントインデックス */
  currentSegmentIndex: number;
  /** 経過時間（ミリ秒） */
  elapsedTime: number;
  /** 総長さ（ミリ秒） */
  totalDuration: number;
}

/** デフォルト設定 */
export const DEFAULT_TEACHER_MODE_SETTINGS: TeacherModeSettings = {
  speed: 1.0,
  pageInterval: 5,
  repeatCount: 1,
  audioQuality: 'standard',
  content: {
    includeTranslation: true,
    includeWordExplanation: true,
    includeGrammarExplanation: false,
    includePronunciationPractice: false,
    includeExampleSentences: false,
  },
};

File: backend/internal/models/teacher_mode.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// TeacherModeSettings は教師モードの設定を保持する
type TeacherModeSettings struct {
	Speed                float64               `json:"speed"`                  // 再生速度 (0.5, 0.75, 1.0, 1.25, 1.5, 2.0)
	PageInterval         int                   `json:"page_interval"`          // ページ間隔（秒）
	RepeatCount          int                   `json:"repeat_count"`           // リピート回数 (1, 2, 3)
	AudioQuality         string                `json:"audio_quality"`          // 音質 ("standard", "premium")
	Content              TeacherModeContent    `json:"content"`                // 学習内容設定
}

// TeacherModeContent は教師モードの学習内容設定を保持する
type TeacherModeContent struct {
	IncludeTranslation           bool `json:"include_translation"`            // 母国語訳を含む
	IncludeWordExplanation       bool `json:"include_word_explanation"`       // 単語解説を含む
	IncludeGrammarExplanation    bool `json:"include_grammar_explanation"`    // 文法解説を含む
	IncludePronunciationPractice bool `json:"include_pronunciation_practice"` // 発音練習を含む
	IncludeExampleSentences      bool `json:"include_example_sentences"`      // 例文を含む
}

// AudioSegmentType は音声セグメントのタイプ
type AudioSegmentType string

const (
	AudioSegmentTypePhrase      AudioSegmentType = "phrase"      // フレーズ
	AudioSegmentTypeTranslation AudioSegmentType = "translation" // 翻訳
	AudioSegmentTypeExplanation AudioSegmentType = "explanation" // 解説
	AudioSegmentTypePause       AudioSegmentType = "pause"       // 一時停止
)

// AudioSegment は音声セグメントを表す
type AudioSegment struct {
	ID       string           `json:"id"`        // セグメントID
	Type     AudioSegmentType `json:"type"`      // タイプ
	AudioURL string           `json:"audio_url"` // 音声URL
	Duration int              `json:"duration"`  // 長さ（ミリ秒）
	Text     string           `json:"text"`      // テキスト
	Language string           `json:"language"`  // 言語
}

// PageAudio はページの音声情報を表す
type PageAudio struct {
	PageNumber    int            `json:"page_number"`    // ページ番号
	Segments      []AudioSegment `json:"segments"`       // セグメント
	TotalDuration int            `json:"total_duration"` // 総長さ（ミリ秒）
}

// TeacherModePlaylist は教師モードのプレイリストを表す
type TeacherModePlaylist struct {
	ID            string              `json:"id"`             // プレイリストID
	BookID        uuid.UUID           `json:"book_id"`        // 書籍ID
	Pages         []PageAudio         `json:"pages"`          // ページ
	Settings      TeacherModeSettings `json:"settings"`       // 設定
	TotalDuration int                 `json:"total_duration"` // 総長さ（ミリ秒）
}

// PlaybackStatus は再生状態
type PlaybackStatus string

const (
	PlaybackStatusStopped PlaybackStatus = "stopped" // 停止
	PlaybackStatusPlaying PlaybackStatus = "playing" // 再生中
	PlaybackStatusPaused  PlaybackStatus = "paused"  // 一時停止
)

// PlaybackState は再生状態を表す
type PlaybackState struct {
	Status               PlaybackStatus `json:"status"`                  // 状態
	CurrentPage          int            `json:"current_page"`            // 現在のページ
	CurrentSegmentIndex  int            `json:"current_segment_index"`   // 現在のセグメントインデックス
	ElapsedTime          int            `json:"elapsed_time"`            // 経過時間（ミリ秒）
	TotalDuration        int            `json:"total_duration"`          // 総長さ（ミリ秒）
}

// TeacherModeDownload は教師モードのダウンロード履歴を表す
type TeacherModeDownload struct {
	ID             uuid.UUID           `json:"id" db:"id"`
	UserID         uuid.UUID           `json:"user_id" db:"user_id"`
	BookID         uuid.UUID           `json:"book_id" db:"book_id"`
	Settings       TeacherModeSettings `json:"settings" db:"settings"`           // JSONBとして保存
	TotalSizeBytes int64               `json:"total_size_bytes" db:"total_size_bytes"`
	DownloadedAt   time.Time           `json:"downloaded_at" db:"downloaded_at"`
	ExpiresAt      *time.Time          `json:"expires_at,omitempty" db:"expires_at"`
	CreatedAt      time.Time           `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time           `json:"updated_at" db:"updated_at"`
}

// TeacherModePlaybackHistory は教師モードの再生履歴を表す
type TeacherModePlaybackHistory struct {
	ID                   uuid.UUID `json:"id" db:"id"`
	UserID               uuid.UUID `json:"user_id" db:"user_id"`
	BookID               uuid.UUID `json:"book_id" db:"book_id"`
	CurrentPage          int       `json:"current_page" db:"current_page"`
	CurrentSegmentIndex  int       `json:"current_segment_index" db:"current_segment_index"`
	ElapsedTime          int       `json:"elapsed_time" db:"elapsed_time"` // 秒単位
	TotalPlayTimeSeconds int       `json:"total_play_time_seconds" db:"total_play_time_seconds"`
	LastPlayedAt         time.Time `json:"last_played_at" db:"last_played_at"`
	CreatedAt            time.Time `json:"created_at" db:"created_at"`
	UpdatedAt            time.Time `json:"updated_at" db:"updated_at"`
}

// PageRange はページ範囲を表す
type PageRange struct {
	Start int `json:"start"` // 開始ページ
	End   int `json:"end"`   // 終了ページ
}
